lib/Sema/CSDiag.cpp

```diff
@@ -5421,8 +5431,92 @@ bool FailureDiagnosis::visitApplyExpr(ApplyExpr *callExpr) {
     // Type check the function subexpression to resolve a type for it if
     // possible.
     fnExpr = typeCheckChildIndependently(callExpr->getFn());
-    if (!fnExpr)
+    if (!fnExpr) {
+      auto UDE = dyn_cast<UnresolvedDotExpr>(callExpr->getFn());
+      if (!UDE)
+        return true;
+
+      auto baseExpr = UDE->getBase();
+      // Offer fix-it for using `subscript` instead of the operator.
+      // FIXME: This doesn't check generic parameters
+      if (baseExpr && UDE->getName().getBaseName() == "subscript") {
+        auto baseType = CS.getType(baseExpr);
+        // Look up subscript declarations and return if there are none.
+        auto lookup = CS.lookupMember(baseType->getRValueType(),
+                                      DeclName(DeclBaseName::createSubscript()));
+        if (lookup.empty())
+          return true;
```

Comment:
We can only return `true` in case there was a diagnostic produced, otherwise we'd end up failing asserts in `ASTVerifier` or in release mode miscompiling or failing in SILGen, if the early exit is required it would be better to return `false` or `CS.TC.Diags.hadAnyError();` and let the result of the diagnostic engine run.